azext_iot/tests/iothub/__init__.py

```diff
@@ -186,6 +182,92 @@ def generate_job_names(self, count=1):
             self.create_random_name(prefix=PREFIX_JOB, length=32) for i in range(count)
         ]
 
+    def _create_storage_account(self):
+        """
+        Create a storage account and container if a storage account was not created yet.
+        Populate the following variables if needed:
+          - storage_account_name
+          - storage_container
+          - storage_cstring
+        """
+        self.entity_rg = ENTITY_RG
+        self.storage_account_name = STORAGE_ACCOUNT_NAME
+        self.storage_container = STORAGE_CONTAINER
+        self.storage_cstring = None
+        if not settings.env.azext_iot_teststorageaccount:
+            storage_list = self.cmd(
+                'storage account list -g "{}"'.format(self.entity_rg)
+            ).get_output_in_json()
+
+            target_storage = None
+            for storage in storage_list:
+                if storage["name"] == self.storage_account_name:
+                    target_storage = storage
+                    break
+
+            if not target_storage:
+                self.cmd(
+                    "storage account create -n {} -g {} --tags iothub={}".format(
+                        self.storage_account_name, self.entity_rg, self.entity_name
+                    )
+                )
+            self._populate_storage_cstring()
+
+        # This won't overwrite or error if container is already created.
+        self.cmd(
+            "storage container create -n {} --connection-string '{}'".format(
+                self.storage_container, self.storage_cstring
+            ),
+        )
+
+    def _populate_storage_cstring(self):
+        """
+        Method to populate storage_cstring
+        """
+        if not self.storage_cstring:
+            self.storage_cstring = self.cmd(
+                "storage account show-connection-string -n {} -g {}".format(
+                    self.storage_account_name, self.entity_rg
+                )
+            ).get_output_in_json()["connectionString"]
+
+    def _delete_storage_account(self):
+        """
+        Delete the storage account if it was created.
+        """
+        if not settings.env.azext_iot_teststorageaccount:
+            self.cmd(
+                "storage account delete -n {} -g {} -y".format(
+                    self.storage_account_name, self.entity_rg
+                ),
+            )
+
+        elif not settings.env.azext_iot_teststorageaccount:
+            self.cmd(
+                "storage container delete -n {} --connection-string '{}'".format(
+                    self.storage_account_name, self.storage_cstring
+                ),
+            )
+
+    def _add_test_tag(self, test_tag):
+        tags = self.cmd(
+            "iot hub show -n {} -g {}".format(self.entity_name, self.entity_rg)
+        ).get_output_in_json()["tags"]
+
+        if tags.get(test_tag):
+            tags[test_tag] = int(tags[test_tag]) + 1
```

Comment:
What are the scenarios when the same test case is run twice where this increment happens?

Reply:
Two major scenarios:
1. test reruns (in the pipeline)
2. providing your own resource

Example of central resource
![image](https://user-images.githubusercontent.com/73560279/162258844-0dcfa3da-ed65-4f70-8c31-185da78c3a9a.png)

Reply:
Isn't this the same single scenario? When a test is re-run against an existing resource? If the pipeline is re-run when no prior created resource is used it should provision new ones.

Can you please explain how adding tags incrementing the number of times a test case is run against an resource helps debugging issues? Is there anything being asserted? Is there a strong correlation between an increment and a pipeline run?  The pipeline logs already show you what tests were run & when....a stronger correlation between the pipeline and the resource may be tagging the pipeline run Id.

Reply:
The tagging should be an extra way to map resource to test in a pipeline. The current set up does not spit out information if a test run is successful and even when it does, it can take a while to find the generated resource name (and sometimes it's not possible to find all generated names). Specifically, this helped with dps test runs where the hubs were in activatedfailed states and the tests did not fail but the hubs were not deleted.

It helps with identifying testing bugs - like when a resource is not deleted correctly (b/c service issue like the activatedfailed hubs or test got cancelled/stopped or some other funky issue). The tests do check that if the autogenerated name is already used then that resource will be used instead. In a perfect world all the test runs should be 1 but I have seen some tests go to two for some reason in a single generated resource- give me a bit to see if I can still recreate.

Having the pipeline log as a tag is a great idea - it would require an extra optional variable. Going to do that. Another idea is to also add the python version in (since the pipelines run 4 times) - but I am not sure on how to do that yet.

Reply:
I also thought about the python version, but applying specifically the pipeline job Id as a tag would already link the resource to a Python version since the IT jobs are organized by Python version.